pkg/cmd/server/api/v1/types.go

```diff
@@ -246,6 +250,31 @@ type MasterConfig struct {
 	AuditConfig AuditConfig `json:"auditConfig"`
 }
 
+// MasterAuthConfig configures authentication options in addition to the standard
+// oauth token and client certificate authenticators
+type MasterAuthConfig struct {
+	RequestHeader *RequestHeaderAuthenticationOptions `json:"requestHeader"`
+}
+
+// RequestHeaderAuthenticationOptions provides options for setting up a front proxy against the entire
+// API instead of against the /oauth endpoint.
+type RequestHeaderAuthenticationOptions struct {
+	// ClientCA is a file with the trusted signer certs.  If empty, no request verification is done, and any direct request to the OAuth server can impersonate any identity from this provider, merely by setting a request header.
+	ClientCA string `json:"clientCA"`
+	// ClientCommonNames is an optional list of common names to require a match from. If empty, any client certificate validated against the clientCA bundle is considered authoritative.
+	ClientCommonNames []string `json:"clientCommonNames"`
+
+	// UsernameHeaders is the list of headers to check for user information.  First hit wins.
+	// X-Remote-User is the default
```

Comment:
> No defaults, right?

Isn't this the default 80+% of the time?  We can distinguish between `nil` and `[]string` (or at least we *should* be able to), why make them specify things we want as default.